File: auth/app/src/main/java/com/google/firebase/quickstart/auth/java/MultiFactorActivity.java
/**
 * Copyright 2016 Google Inc. All Rights Reserved.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.firebase.quickstart.auth.java;

import android.app.AlertDialog;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.MultiFactorInfo;
import com.google.firebase.auth.PhoneMultiFactorInfo;
import com.google.firebase.quickstart.auth.R;
import com.google.firebase.quickstart.auth.databinding.ActivityMultiFactorBinding;

import java.util.List;

public class MultiFactorActivity extends BaseActivity implements
        View.OnClickListener {

    public static final int RESULT_NEEDS_MFA_SIGN_IN = 42;

    private static final String TAG = "MultiFactor";
    private static final int RC_MULTI_FACTOR = 9005;

    private ActivityMultiFactorBinding mBinding;

    // [START declare_auth]
    private FirebaseAuth mAuth;
    // [END declare_auth]

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mBinding = ActivityMultiFactorBinding.inflate(getLayoutInflater());
        setContentView(mBinding.getRoot());
        setProgressBar(mBinding.progressBar);

        // Buttons
        mBinding.emailSignInButton.setOnClickListener(this);
        mBinding.signOutButton.setOnClickListener(this);
        mBinding.verifyEmailButton.setOnClickListener(this);
        mBinding.enrollMfa.setOnClickListener(this);
        mBinding.unenrollMfa.setOnClickListener(this);
        mBinding.reloadButton.setOnClickListener(this);

        // [START initialize_auth]
        // Initialize Firebase Auth
        mAuth = FirebaseAuth.getInstance();
        // [END initialize_auth]

        showDisclaimer();
    }

    // [START on_start_check_user]
    @Override
    public void onStart() {
        super.onStart();
        // Check if user is signed in (non-null) and update UI accordingly.
        FirebaseUser currentUser = mAuth.getCurrentUser();
        updateUI(currentUser);
    }
    // [END on_start_check_user]


    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == RC_MULTI_FACTOR) {
            if (resultCode == RESULT_NEEDS_MFA_SIGN_IN) {
                Intent intent = new Intent(this, MultiFactorSignInActivity.class);
                intent.putExtras(data.getExtras());
                startActivityForResult(intent, RC_MULTI_FACTOR);
            }
        }
    }

    private void signOut() {
        mAuth.signOut();
        updateUI(null);
    }

    private void sendEmailVerification() {
        // Disable button
        mBinding.verifyEmailButton.setEnabled(false);

        // Send verification email
        // [START send_email_verification]
        final FirebaseUser user = mAuth.getCurrentUser();
        user.sendEmailVerification()
                .addOnCompleteListener(this, new OnCompleteListener<Void>() {
                    @Override
                    public void onComplete(@NonNull Task<Void> task) {
                        // [START_EXCLUDE]
                        // Re-enable button
                        mBinding.verifyEmailButton.setEnabled(true);

                        if (task.isSuccessful()) {
                            Toast.makeText(MultiFactorActivity.this,
                                    "Verification email sent to " + user.getEmail(),
                                    Toast.LENGTH_SHORT).show();
                        } else {
                            Log.e(TAG, "sendEmailVerification", task.getException());
                            Toast.makeText(MultiFactorActivity.this,
                                    "Failed to send verification email.",
                                    Toast.LENGTH_SHORT).show();
                        }
                        // [END_EXCLUDE]
                    }
                });
        // [END send_email_verification]
    }

    private void reload() {
        mAuth.getCurrentUser().reload().addOnCompleteListener(new OnCompleteListener<Void>() {
            @Override
            public void onComplete(@NonNull Task<Void> task) {
                if (task.isSuccessful()) {
                    updateUI(mAuth.getCurrentUser());
                    Toast.makeText(MultiFactorActivity.this,
                            "Reload successful!",
                            Toast.LENGTH_SHORT).show();
                } else {
                    Log.e(TAG, "reload", task.getException());
                    Toast.makeText(MultiFactorActivity.this,
                            "Failed to reload user.",
                            Toast.LENGTH_SHORT).show();
                }
            }
        });
    }

    private void updateUI(FirebaseUser user) {
        hideProgressBar();
        if (user != null) {
            mBinding.status.setText(getString(R.string.emailpassword_status_fmt,
                    user.getEmail(), user.isEmailVerified()));
            mBinding.detail.setText(getString(R.string.firebase_status_fmt, user.getUid()));

            List<MultiFactorInfo> secondFactors = user.getMultiFactor().getEnrolledFactors();

            if (secondFactors.isEmpty()) {
                mBinding.unenrollMfa.setVisibility(View.GONE);
            } else {
                mBinding.unenrollMfa.setVisibility(View.VISIBLE);

                StringBuilder sb = new StringBuilder("Second Factors: ");
                String delimiter = ", ";
                for (MultiFactorInfo x : secondFactors) {
                    sb.append(((PhoneMultiFactorInfo) x).getPhoneNumber() + delimiter);
                }
                sb.setLength(sb.length() - delimiter.length());
                mBinding.mfaInfo.setText(sb.toString());
            }

            mBinding.emailPasswordButtons.setVisibility(View.GONE);
            mBinding.signedInButtons.setVisibility(View.VISIBLE);

            int reloadVisibility = secondFactors.isEmpty() ? View.VISIBLE : View.GONE;
            mBinding.reloadButton.setVisibility(reloadVisibility);

            if (user.isEmailVerified()) {
                mBinding.verifyEmailButton.setVisibility(View.GONE);
                mBinding.enrollMfa.setVisibility(View.VISIBLE);
            } else {
                mBinding.verifyEmailButton.setVisibility(View.VISIBLE);
                mBinding.enrollMfa.setVisibility(View.GONE);
            }
        } else {
            mBinding.status.setText(R.string.multi_factor_signed_out);
            mBinding.detail.setText(null);
            mBinding.mfaInfo.setText(null);

            mBinding.emailPasswordButtons.setVisibility(View.VISIBLE);
            mBinding.signedInButtons.setVisibility(View.GONE);
        }
    }

    private void showDisclaimer() {
        new AlertDialog.Builder(this)
                .setTitle("Warning")
                .setMessage("Multi-factor authentication with SMS is currently only available for " +
                        "Google Cloud Identity Platform projects. For more information see: " +
                        "https://cloud.google.com/identity-platform/docs/android/mfa")
                .setPositiveButton("OK", null)
                .show();
    }

    @Override
    public void onClick(View v) {
        int i = v.getId();
        if (i == R.id.emailSignInButton) {
            startActivityForResult(new Intent(this, EmailPasswordActivity.class), RC_MULTI_FACTOR);
        } else if (i == R.id.signOutButton) {
            signOut();
        } else if (i == R.id.verifyEmailButton) {
            sendEmailVerification();
        } else if (i == R.id.enrollMfa) {
            startActivity(new Intent(this, MultiFactorEnrollActivity.class));
        } else if (i == R.id.unenrollMfa) {
            startActivity(new Intent(this, MultiFactorUnenrollActivity.class));
        } else if (i == R.id.reloadButton) {
            reload();
        }
    }
}

File: auth/app/src/main/java/com/google/firebase/quickstart/auth/kotlin/MultiFactorActivity.kt
package com.google.firebase.quickstart.auth.kotlin

import android.app.AlertDialog
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Toast
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.PhoneMultiFactorInfo
import com.google.firebase.quickstart.auth.R
import com.google.firebase.quickstart.auth.databinding.ActivityMultiFactorBinding

class MultiFactorActivity : BaseActivity(), View.OnClickListener {
    // [START declare_auth]
    private lateinit var auth: FirebaseAuth
    // [END declare_auth]

    private lateinit var binding: ActivityMultiFactorBinding

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMultiFactorBinding.inflate(layoutInflater)
        setContentView(binding.reloadButton)
        setProgressBar(binding.progressBar)

        // Buttons
        binding.emailSignInButton.setOnClickListener(this)
        binding.signOutButton.setOnClickListener(this)
        binding.verifyEmailButton.setOnClickListener(this)
        binding.enrollMfa.setOnClickListener(this)
        binding.unenrollMfa.setOnClickListener(this)
        binding.reloadButton.setOnClickListener(this)

        // [START initialize_auth]
        // Initialize Firebase Auth
        auth = FirebaseAuth.getInstance()
        // [END initialize_auth]

        showDisclaimer()
    }

    // [START on_start_check_user]
    public override fun onStart() {
        super.onStart()
        // Check if user is signed in (non-null) and update UI accordingly.
        val currentUser = auth.currentUser
        updateUI(currentUser)
    }

    // [END on_start_check_user]
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == RC_MULTI_FACTOR) {
            if (resultCode == RESULT_NEEDS_MFA_SIGN_IN) {
                val intent = Intent(this, MultiFactorSignInActivity::class.java)
                intent.putExtras(data!!.extras!!)
                startActivityForResult(intent, RC_MULTI_FACTOR)
            }
        }
    }

    private fun signOut() {
        auth.signOut()
        updateUI(null)
    }

    private fun sendEmailVerification() { // Disable button
        findViewById<View>(R.id.verifyEmailButton).isEnabled = false
        // Send verification email
        // [START send_email_verification]
        val user = auth.currentUser
        user!!.sendEmailVerification()
                .addOnCompleteListener(this) { task ->
                    // [START_EXCLUDE]
                    // Re-enable button
                    findViewById<View>(R.id.verifyEmailButton).isEnabled = true
                    if (task.isSuccessful) {
                        Toast.makeText(this@MultiFactorActivity,
                                "Verification email sent to " + user.email,
                                Toast.LENGTH_SHORT).show()
                    } else {
                        Log.e(TAG, "sendEmailVerification", task.exception)
                        Toast.makeText(this@MultiFactorActivity,
                                "Failed to send verification email.",
                                Toast.LENGTH_SHORT).show()
                    }
                    // [END_EXCLUDE]
                }
        // [END send_email_verification]
    }

    private fun reload() {
        auth.currentUser!!.reload().addOnCompleteListener { task ->
            if (task.isSuccessful) {
                updateUI(auth.currentUser)
                Toast.makeText(this@MultiFactorActivity,
                        "Reload successful!",
                        Toast.LENGTH_SHORT).show()
            } else {
                Log.e(TAG, "reload", task.exception)
                Toast.makeText(this@MultiFactorActivity,
                        "Failed to reload user.",
                        Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun updateUI(user: FirebaseUser?) {
        hideProgressBar()
        if (user != null) {
            binding.status.text = getString(R.string.emailpassword_status_fmt,
                    user.email, user.isEmailVerified)
            binding.detail.text = getString(R.string.firebase_status_fmt, user.uid)
            val secondFactors = user.multiFactor.enrolledFactors
            if (secondFactors.isEmpty()) {
                findViewById<View>(R.id.unenrollMfa).visibility = View.GONE
            } else {
                findViewById<View>(R.id.unenrollMfa).visibility = View.VISIBLE
                val sb = StringBuilder("Second Factors: ")
                val delimiter = ", "
                for (x in secondFactors) {
                    sb.append((x as PhoneMultiFactorInfo).phoneNumber + delimiter)
                }
                sb.setLength(sb.length - delimiter.length)
                binding.mfaInfo.text = sb.toString()
            }
            findViewById<View>(R.id.emailPasswordButtons).visibility = View.GONE
            findViewById<View>(R.id.signedInButtons).visibility = View.VISIBLE
            val reloadVisibility = if (secondFactors.isEmpty()) View.VISIBLE else View.GONE
            findViewById<View>(R.id.reloadButton).visibility = reloadVisibility
            if (user.isEmailVerified) {
                findViewById<View>(R.id.verifyEmailButton).visibility = View.GONE
                findViewById<View>(R.id.enrollMfa).visibility = View.VISIBLE
            } else {
                findViewById<View>(R.id.verifyEmailButton).visibility = View.VISIBLE
                findViewById<View>(R.id.enrollMfa).visibility = View.GONE
            }
        } else {
            binding.status.setText(R.string.multi_factor_signed_out)
            binding.detail.text = null
            binding.mfaInfo.text = null
            findViewById<View>(R.id.emailPasswordButtons).visibility = View.VISIBLE
            findViewById<View>(R.id.signedInButtons).visibility = View.GONE
        }
    }

    private fun showDisclaimer() {
        AlertDialog.Builder(this)
                .setTitle("Warning")
                .setMessage("Multi-factor authentication with SMS is currently only available for " +
                        "Google Cloud Identity Platform projects. For more information see: " +
                        "https://cloud.google.com/identity-platform/docs/android/mfa")
                .setPositiveButton("OK", null)
                .show()
    }

    override fun onClick(v: View) {
        val i = v.id
        when (i) {
            R.id.emailSignInButton -> {
                startActivityForResult(Intent(this, EmailPasswordActivity::class.java), RC_MULTI_FACTOR)
            }
            R.id.signOutButton -> signOut()
            R.id.verifyEmailButton -> sendEmailVerification()
            R.id.enrollMfa -> startActivity(Intent(this, MultiFactorEnrollActivity::class.java))
            R.id.unenrollMfa -> startActivity(Intent(this, MultiFactorUnenrollActivity::class.java))
            R.id.reloadButton -> reload()
        }
    }

    companion object {
        const val RESULT_NEEDS_MFA_SIGN_IN = 42
        private const val TAG = "MultiFactor"
        private const val RC_MULTI_FACTOR = 9005
    }
}
